Scripts/Runtime/Entities/Tasks/Jobs/JobTypes/TaskCancelJobForDeferExtension.cs

```diff
@@ -0,0 +1,136 @@
+            [NativeSetThreadIndex] internal readonly int NativeThreadIndex;
+
+            public WrapperJobStruct(ref TJob jobData,
+                                    ref CancelTaskStreamScheduleInfo<TInstance> scheduleInfo)
+            {
+                JobData = jobData;
+                CancellationUpdater = scheduleInfo.CancellationUpdater;
+                NativeThreadIndex = UNSET_NATIVE_THREAD_INDEX;
+            }
+        }
+
+        //*************************************************************************************************************
+        // PRODUCER
+        //*************************************************************************************************************
+        private struct WrapperJobProducer<TJob, TInstance>
+            where TJob : struct, ITaskCancelJobForDefer<TInstance>
+            where TInstance : unmanaged, IEntityProxyInstance
+        {
+            // ReSharper disable once StaticMemberInGenericType
+            internal static readonly IntPtr JOB_REFLECTION_DATA = JobsUtility.CreateJobReflectionData(typeof(WrapperJobStruct<TJob, TInstance>),
+                                                                                                      typeof(TJob),
+                                                                                                      (ExecuteJobFunction)Execute);
+
+
+            private delegate void ExecuteJobFunction(ref WrapperJobStruct<TJob, TInstance> jobData,
+                                                     IntPtr additionalPtr,
+                                                     IntPtr bufferRangePatchData,
+                                                     ref JobRanges ranges,
+                                                     int jobIndex);
+
+
+            [SuppressMessage("ReSharper", "MemberCanBePrivate.Global", Justification = "Required by Burst.")]
+            public static unsafe void Execute(ref WrapperJobStruct<TJob, TInstance> wrapperData,
+                                              IntPtr additionalPtr,
+                                              IntPtr bufferRangePatchData,
+                                              ref JobRanges ranges,
+                                              int jobIndex)
+            {
+                ref TJob jobData = ref wrapperData.JobData;
+                ref DataStreamCancellationUpdater<TInstance> cancellationUpdater = ref wrapperData.CancellationUpdater;
+                
+                cancellationUpdater.InitForThread(wrapperData.NativeThreadIndex);
+                jobData.InitForThread(wrapperData.NativeThreadIndex);
+
+                while (true)
+                {
+                    if (!JobsUtility.GetWorkStealingRange(ref ranges, jobIndex, out int beginIndex, out int endIndex))
+                    {
+                        return;
+                    }
+
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
```

Comment:
I don't think there's a clean way to do that short of adding a ton of overhead for each thread this would run on.

This `while` loop is exactly the same as Unity's for all of their jobs with the exit condition being the `JobsUtility.GetWorkStealingRange` function. That function goes into Unity's C++ land for the job scheduling. For an infinite loop to occur, Unity's internals would need to fail and then nothing is going to work anyway.

Reply:
hmmm, ok. You could have a counter or time stamp initialized in a debug only method outside the loop and then call a debug check method on each iteration for some threshold.

I'm on the fence about trusting Unity's QA....you know? I guess leave it as is and the first time we get bit by a bug like this we implement a fix across all of Anvil.

Reply:
> You could have a counter or time stamp initialized in a debug only method outside the loop and then call a debug check method on each iteration for some threshold.

Yeah that could work. Would just have to use some timer mechanism that doesn't sync on any other thread.

> I'm on the fence about trusting Unity's QA....you know? I guess leave it as is and the first time we get bit by a bug like this we implement a fix across all of Anvil.

Hahah yes, agreed. I've added a low priority TODO. #85